<!DOCTYPE html>
    <style>
        table.mytable-marg{
            border-collapse: collapse;
        }
        table.mytable-marg td, table.mytable-marg th{
          border: 1px solid #ccc;
          text-align: left;
          padding-right: 18px; 
          padding-left: 10px;
          font-weight: normal
        }
        .inline-2perline{
            display: inline-block;   
            width: 49%;
        }
        .full-div {
            vertical-align: top
            margin-left: 8px;
            margin-right: 8px;
        }
        .verdanafont {
            font-family:'verdana'
        }
        h100 { font-family:"Courier New"; display:inline }
    </style>
    <div class="verdanafont">
    

<h1> Trimviz trimming summary: Fastq-bam soft-clipping mode </h1> <br><hr/><br><div>

<h3> Input files: </h3>
-P (input R2 fastq file): <h100> ../tv_data/post_trimmed_R2.fastq.gz </h100><br>
-b (bam file): <h100> ../tv_data/alignment.bam </h100><br>
-g (genome fasta file): <h100> ../tv_data/genome.fa </h100><br></div><br><hr/>

<h3> Summary of trimming frequency </h3> 
 <table class="mytable-marg" ><tr><th><b>Trim class</b></th><th><b>Number of reads</b></th><th><b> % </b></th></tr>
<tr><th> uncut </th><th> 6687 </th><th> 13.4 </th>
<tr><th> 5pcut </th><th> 2505 </th><th> 5.0 </th>
<tr><th> 3pcut </th><th> 41250 </th><th> 82.5 </th>
<tr><th> removed </th><th> 882 </th><th> 1.8 </th>
<tr><th> generated_warning </th><th> 0 </th><th> 0.0 </th>
<tr><th> indel </th><th> 1249 </th><th> 2.5 </th>
<tr><th> ambiguous_cut </th><th> 0 </th><th> 0.0 </th>
<tr><th><b>Tot unique</b></th><th><b>50000</b></th><th></th></tr>
</table><br><hr/>

<div class="inline-2perline">
3p trim position profile: <br><embed src= profile_3pcut.pdf type="application/pdf" width="100%" height="650px" /><br>
</div><div class="inline-2perline">
5p trim position profile: <br><embed src= profile_5pcut.pdf type="application/pdf" width="100%" height="650px" /><br>
</div><br><hr/><div>

<hr/><details> <summary> Heatmap, 3'-trimmed reads clustered by untrimmed read sequence </summary> <br>
<embed src= TVheatmap_S_3p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Heatmap, 3'-trimmed reads clustered by quality patterns </summary> <br>
<embed src= TVheatmap_Q_3p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Heatmaps, 3'-trimmed reads clustered by local reference sequence </summary> <br>
<embed src= TVheatmap_G_3p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Heatmap, 5'-trimmed reads clustered by untrimmed read sequence </summary> <br>
<embed src= TVheatmap_S_5p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Heatmap, 5'-trimmed reads clustered by quality patterns </summary> <br>
<embed src= TVheatmap_Q_5p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Heatmaps, 5'-trimmed reads clustered by local reference sequence </summary> <br>
<embed src= TVheatmap_G_5p.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

<hr/><details> <summary> Individual read trimming, grouped by trimming class </summary> <br>
<embed src= indiv_reads.pdf type="application/pdf" width="100%" height="600px" />
<br></details><br>

</div><hr/></div><br><br>
